### DISCUSSION TITLE
Fatal TweakScale Error

### DISCUSSION THREAD

**Discussion by Unavailable-is-Username** [TweakScale/TweakScale#306] [Support] [answered]
@Lisias
I honestly have zero idea what is causing this error to occur, despite (at least believing I am) being fairly good at diagnosing issues. 

https://1drv.ms/f/s!AlqfjkpGLMMhjIUTKIqY471w5xuKtw?e=9kJ4CT

The link I provide to you is from my OneDrive and contains:
- A complete archive of my entire 14.5GB KSP folder (in 7zip Format). Included as I believe the issue/s could be related to things outside the usual Gamedata folder (and to give you a 1:1 copy of my situation)
- A Screenshot of my Gamedata folder
- The KSP.log
- A Text file listing all the installed mods that state they do not support the version of KSP I am running. (Although, I have not personally had any issues with them)
- A text file listing all the mods installed (taken from the KSP.log)

**Comment by Lisias** [comment]
@Unavailable-is-Username ,

> I honestly have zero idea what is causing this error to occur, despite (at least believing I am) being fairly good at diagnosing issues.

That makes two of us! :P  Let's fix this. :)

I'm downloading the bomb. Usually only the `KSP.log` is more than enough, because it's pretty rare something blowing up without the log telling us what happened.

I will come back to you in about an hour (house keeping day).

**Comment by Lisias** [comment]
@Unavailable-is-Username , found the `KSP.log` alone on the folder and inspected it. I think the problem is EnvironmentalVisualEnhancements  and Kopernicus:

```
[LOG 20:37:54.708] Load(Assembly): EnvironmentalVisualEnhancements/Plugins/_BuildManager
<….>
[LOG 20:37:56.201] [AddonLoader]: Instantiating addon '_BuildManager' from assembly '_BuildManager'
[EXC 20:37:56.210] NotSupportedException: The invoked member is not supported in a dynamic module.
        System.Reflection.Emit.AssemblyBuilder.get_Location () (at <9577ac7a62ef43179789031239ba8798>:0)
        _BuildManager._BuildManager+<>c__DisplayClass2_0.<logVersion>b__0 (System.Reflection.Assembly x) (at <5329a27ec3b64178b009f7e4fdd9f215>:0)
        System.Linq.Enumerable+WhereSelectArrayIterator`2[TSource,TResult].ToList () (at <351e49e2a5bf4fd6beabb458ce2255f3>:0)
        System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) (at <351e49e2a5bf4fd6beabb458ce2255f3>:0)
        _BuildManager._BuildManager.logVersion () (at <5329a27ec3b64178b009f7e4fdd9f215>:0)
        _BuildManager._BuildManager.Awake () (at <5329a27ec3b64178b009f7e4fdd9f215>:0)
        UnityEngine.DebugLogHandler:LogException(Exception, Object)
        KSPe.Util.Log.UnityLogDecorator:UnityEngine.ILogHandler.LogException(Exception, Object)
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
        UnityEngine.GameObject:AddComponent(Type)
        AddonLoader:StartAddon(LoadedAssembly, Type, KSPAddon, Startup)
        AddonLoader:StartAddons(Startup)
        <LoadObjects>d__90:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        <CreateDatabase>d__71:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        GameDatabase:StartLoad()
        <LoadSystems>d__11:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        LoadingScreen:Start()
```

and

```
[LOG 20:37:56.557] [AddonLoader]: Instantiating addon 'PQSModCreator' from assembly 'KopernicusExpansion.RegionalPQSMods'
[EXC 20:37:56.569] ReflectionTypeLoadException: Exception of type 'System.Reflection.ReflectionTypeLoadException' was thrown.
        System.Reflection.Assembly.GetTypes () (at <9577ac7a62ef43179789031239ba8798>:0)
        KopernicusExpansion.RegionalPQSMods.PQSModCreator+<>c.<GetModTypes>b__0_1 (System.Reflection.Assembly a) (at <f5e1172978524a599c9056aa1686da11>:0)
        System.Linq.Enumerable+SelectManySingleSelectorIterator`2[TSource,TResult].MoveNext () (at <351e49e2a5bf4fd6beabb458ce2255f3>:0)
        KopernicusExpansion.RegionalPQSMods.PQSModCreator.GetModTypes () (at <f5e1172978524a599c9056aa1686da11>:0)
        KopernicusExpansion.RegionalPQSMods.PQSModCreator.Awake () (at <f5e1172978524a599c9056aa1686da11>:0)
        UnityEngine.DebugLogHandler:LogException(Exception, Object)
        KSPe.Util.Log.UnityLogDecorator:UnityEngine.ILogHandler.LogException(Exception, Object)
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
        UnityEngine.GameObject:AddComponent(Type)
        AddonLoader:StartAddon(LoadedAssembly, Type, KSPAddon, Startup)
        AddonLoader:StartAddons(Startup)
        <LoadObjects>d__90:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        <CreateDatabase>d__71:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        GameDatabase:StartLoad()
        <LoadSystems>d__11:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        LoadingScreen:Start()
```

One of these two (or both) are triggering the infamous `Assembly Loader/Resolver` bug on KSP, and when this happens everything and the kitchen's sink borks when trying to use a thingy called Reflection or loading an Assembly - and this includes KSP itself.

Remove both (EVE and Kopernicus) and see if the problem goes away. If yes, reinstall EVE and see if the problem is back. If not, remove it again and install Kopernicus, and see if the problem goes away.

The one that borks again is the troublemaker, and then you need to reach the respective maintainer for help. I think I remember that Kopernicus had a mishap on dependencies earlier this year, so perhaps by just updating Kopernicus to the latest will save the day (remember to install the dependencies).

**Reply by Unavailable-is-Username** [reply]
@Lisias,
I tried removing EVE first since Kopernicus is a dependency of Kerbal Galaxy and I know from personal experience that messing with either of those will cause other problems (namely savegame issues). Tried twice and twice KSP terminated before getting to the loading screen. (I have updated the drive with a new KSP.log file) 

Having both EVE and Kopernicus (and its related mods) removed caused the same result. However, adding EVE back in has allowed KSP to load. Which only further perplexes me. The game did load but Tweakscale did give me a fatal error upon reaching the main menu (included is an image of the error). I feel like this clears Kopernicus and Kerbal Galaxy as being the culprits here. It is likely they are merely victims of that loader bug. But it does make me suspicous of EVE.
I'll update the drive with a new log (I am naming these btw, not overwriting them).


Also, another thing I should note. Tweakscale has only shown the error after loading into the main menu, in my experience that doesn't happen often.

**Reply by Lisias** [reply]
> Also, another thing I should note. Tweakscale has only shown the error after loading into the main menu, in my experience that doesn't happen often.

These errors are runtime errors (like bad patching) detected by TweakScale, if you get them KSP itself is ok and the problem are 3rd parties screwing up patches.

Let's fix a problem at time, these ones that happens on Main Menu are secondary ones.

**Comment by Lisias** [comment]
@Unavailable-is-Username , I found something!

I just detected that TweakScake Companion for Frameworks is still at relase 0.2.0.0! Please updated it to the latest:

https://github.com/TweakScale/Companion_Frameworks/releases

This will fix at leat **one** issue I just found on the latest KSP.log you sent!

**Reply by Unavailable-is-Username** [reply]
Alright, I'll install the latest Uber pack and test it.

**Comment by Lisias** [comment]
About the FATALity on TweaskScale, the problem is BDArmory:

```
[LOG 07:41:04.534] [TweakScale] ERROR: **FATAL** Part bdWarheadSmall (Small High Explosive Warhead) has a fatal problem due having duplicated properties - see issue [#34
```

I found the culprit on BDA itself:

```
[LOG 07:35:28.897] [ModuleManager] INFO: Applying update BDArmory/MMPatches/BDArmory_TweakScale /@PART[bdWarheadSmall] to BDArmory/Parts/smallWarhead/part.cfg/PART[bdWarheadSmall]
[LOG 07:35:29.328] [ModuleManager] INFO: Applying update BDArmory/MMPatches/BDA_TweakScale/@PART[bdWarheadSmall]:NEEDS[TweakScale] to BDArmory/Parts/smallWarhead/part.cfg/PART[bdWarheadSmall]
```

Please reach the BDA maintainers and ask them what of the patch above need to be removed!

**Reply by Unavailable-is-Username** [reply]
It might have been caused by me updating from BDA to BDAC I'll delete BDAC and see if BDA+ Solves the issue.

**Comment by Unavailable-is-Username** [comment]
Okay... I have good and bad news.

Good news - installing TWC and BDA+ seems to have resolved issues with Tweakscale.

Bad News - Module Manager is still giving me a RLTX error (which I have been ignoring so far) and Kopernicus is giving me what I'll assume is its own fatal error

(New log uploaded if you want to look at it)

**Comment by Lisias** [comment]
Well, let's see Kopernicus:

```
[EXC 16:35:00.380] ReflectionTypeLoadException: Exception of type 'System.Reflection.ReflectionTypeLoadException' was thrown.
        System.Reflection.Assembly.GetTypes () (at <9577ac7a62ef43179789031239ba8798>:0)
        Kopernicus.ConfigParser.Parser+<>c.<GetModTypes>b__16_1 (System.Reflection.Assembly a) (at <b81c7e09ce34479e9d3831f2ed50b754>:0)
        System.Linq.Enumerable+SelectManySingleSelectorIterator`2[TSource,TResult].MoveNext () (at <351e49e2a5bf4fd6beabb458ce2255f3>:0)
        Kopernicus.ConfigParser.Parser.GetModTypes () (at <b81c7e09ce34479e9d3831f2ed50b754>:0)
        Kopernicus.ConfigParser.Parser.get_ModTypes () (at <b81c7e09ce34479e9d3831f2ed50b754>:0)
        Kopernicus.RuntimeUtility.LogAggregatorWorker.Awake () (at <ed5211404d034a8985de0866662fb372>:0)
        UnityEngine.DebugLogHandler:LogException(Exception, Object)
        KSPe.Util.Log.UnityLogDecorator:UnityEngine.ILogHandler.LogException(Exception, Object)
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
        UnityEngine.GameObject:AddComponent(Type)
        AddonLoader:StartAddon(LoadedAssembly, Type, KSPAddon, Startup)
        AddonLoader:StartAddons(Startup)
        <LoadObjects>d__90:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        <CreateDatabase>d__71:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        GameDatabase:StartLoad()
        <LoadSystems>d__11:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        LoadingScreen:Start()
```

```
[EXC 16:35:00.436] ReflectionTypeLoadException: Exception of type 'System.Reflection.ReflectionTypeLoadException' was thrown.
        System.Reflection.Assembly.GetTypes () (at <9577ac7a62ef43179789031239ba8798>:0)
        KopernicusExpansion.RegionalPQSMods.PQSModCreator+<>c.<GetModTypes>b__0_1 (System.Reflection.Assembly a) (at <f5e1172978524a599c9056aa1686d
        System.Linq.Enumerable+SelectManySingleSelectorIterator`2[TSource,TResult].MoveNext () (at <351e49e2a5bf4fd6beabb458ce2255f3>:0)
        KopernicusExpansion.RegionalPQSMods.PQSModCreator.GetModTypes () (at <f5e1172978524a599c9056aa1686da11>:0)
        KopernicusExpansion.RegionalPQSMods.PQSModCreator.Awake () (at <f5e1172978524a599c9056aa1686da11>:0)
        UnityEngine.DebugLogHandler:LogException(Exception, Object)
        KSPe.Util.Log.UnityLogDecorator:UnityEngine.ILogHandler.LogException(Exception, Object)
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
        UnityEngine.GameObject:AddComponent(Type)
        AddonLoader:StartAddon(LoadedAssembly, Type, KSPAddon, Startup)
        AddonLoader:StartAddons(Startup)
        <LoadObjects>d__90:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        <CreateDatabase>d__71:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        GameDatabase:StartLoad()
        <LoadSystems>d__11:MoveNext()
        UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
        LoadingScreen:Start()
```

and I just found another one!!

```
[WRN 16:35:00.728] [KSPCF] A ReflectionTypeLoadException thrown by Assembly.GetTypes() has been handled by KSP Community Fixes.
This is usually harmless, but indicates that the "OneTimeResourceConverter" plugin failed to load (location: "GameData\GroundConstruction\Plugins\OneTimeResourceConverter.dll")
Stacktrace:
  at AssemblyLoader+LoadedAssembyList.GetPathByType (System.Type type) [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at AssemblyLoader.GetPathByType (System.Type type) [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at KSP.IO.IOUtils.GetFilePathFor (System.Type T, System.String file, Vessel flight) [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at KSP.IO.PluginConfiguration.CreateForType[T] (Vessel flight) [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at MPUtils.MPConfig.Getconfig () [0x00000] in <4b14005524b04aca9e78cfaf2c6b44ee>:0
  at MPUtils.MPStartup.Awake () [0x00000] in <4b14005524b04aca9e78cfaf2c6b44ee>:0
  at UnityEngine.GameObject.Internal_AddComponentWithType (System.Type componentType) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at UnityEngine.GameObject.AddComponent (System.Type componentType) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at AddonLoader.StartAddon (AssemblyLoader+LoadedAssembly asm, System.Type type, KSPAddon addon, KSPAddon+Startup level) [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at AddonLoader.StartAddons (KSPAddon+Startup level) [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at GameDatabase+<LoadObjects>d__90.MoveNext () [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at UnityEngine.MonoBehaviour.StartCoroutineManaged2 (System.Collections.IEnumerator enumerator) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at UnityEngine.MonoBehaviour.StartCoroutine (System.Collections.IEnumerator routine) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at GameDatabase+<CreateDatabase>d__71.MoveNext () [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at UnityEngine.MonoBehaviour.StartCoroutineManaged2 (System.Collections.IEnumerator enumerator) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at UnityEngine.MonoBehaviour.StartCoroutine (System.Collections.IEnumerator routine) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at GameDatabase.StartLoad () [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at LoadingScreen+<LoadSystems>d__11.MoveNext () [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at UnityEngine.MonoBehaviour.StartCoroutineManaged2 (System.Collections.IEnumerator enumerator) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at UnityEngine.MonoBehaviour.StartCoroutine (System.Collections.IEnumerator routine) [0x00000] in <12e76cd50cc64cf19e759e981cb725af>:0
  at LoadingScreen.Start () [0x00000] in <4b449f2841f84227adfaad3149c8fdba>:0
```

This one is from Ground Construction! This one is being flagged by a lot of 3rd parties.

What I don't know if this Ground Construction is the real culprit, and Kopernicus is borking by side effect (that `Assembly Loader/Resolver` crap), or vice versa.

Remove Ground Construction and see if the situation improves. Otherwise, it's something fishy on Kopernicus.

In time, your Kopernicus is updated to the latest version? Perhaps we are fighting something that it's already fixed… Same for Ground Construction!

**Reply by Unavailable-is-Username** [reply]
Given that Ground Construction is a dependency of another mod, I opted to remove OneTimeResourceConverter.dll and see if that works. Although it took two runs. The second run showed no errors. I have uploaded a new log. If you could look through it, that would be great. (There may be errors the game isn't yet making me aware of, nor am I sure how the removal of that dll has affected things) Also, I updated Kopernicus yesterday to v175.

**Reply by Unavailable-is-Username** [reply]
![image](https://github.com/TweakScale/TweakScale/assets/64349888/6f1e89fd-9558-47c1-94b5-acba65e2a168)
![image](https://github.com/TweakScale/TweakScale/assets/64349888/da5a94c5-8d3e-4221-90b7-4bb1c32bf5ab)

I loaded a savegame and this is the first time I've seen something like this. I hope it is something that is supposed to happen.

**Comment by Lisias** [comment, marked as answer]
I can't say if it's good or not, because I really don't know Kopernicus neither Ground Construction…

But since removing that DLL from GC fixed the problem, well… We found the root cause of it.

IMHO, you should reach Ground Construction for help - I don't know if removing this DLL would not cause side effects later!

P.S.: Hey. you are using Impossible Innovations!! :) Now and then I throw plausibility through the Windows and toy with it, it's fun! :)

**Reply by Unavailable-is-Username** [reply]
Hehe It is quite fun to mess around with :)
Thank you for your help Lisias, you have been fantastic!

P:S @Lisias I uploaded a new log if you want to look into it (taken after savegame was loaded). As I am not fully convinced that GC could cause this, given I have FARC installed
